File: sdks/node/examples/sidecar.ts
// openrate SIDECAR mode on Node — the `openrate` binary in a child process,
// HTTP over loopback.
//
//   npm run build && node examples/sidecar.ts
//
// No FFI, no native dependency, no Go runtime in this process, and no fork
// hazard. It needs the binary — set OPENRATE_BINARY, or build one:
//   go build -o /tmp/openrate ../../cmd/openrate
//
// A server refreshes on startup, so this example DOES use the network. If the
// machine is offline it reports WHICH source failed and how, then exits 0
// rather than pretending. To watch that path without unplugging anything:
//   HTTPS_PROXY=http://127.0.0.1:1 OPENRATE_READY_TIMEOUT_MS=3000 node examples/sidecar.ts

import { Sidecar } from "../index.js";

interface Convert {
  from: string;
  to: string;
  amount: number;
  result: number;
  rate: { path: string[]; sources: string[]; hops: number; age_sec: number; quality: { grade: string } };
}

console.log(`node       ${process.version} on ${process.platform}/${process.arch}`);

// -ui=false leaves only the JSON API. -sources ecb keeps the example to one
// upstream so a slow source cannot stretch it out.
const side = await Sidecar.start({ base: "ZAR", sources: "ecb", ui: false });
try {
  console.log(`sidecar    ${side.baseURL}`);
  console.log(`api        ${side.apiBaseURL}\n`);

  // waitForRates polls /readyz on a Node timer — the event loop is never
  // blocked, which is the whole difference from direct mode's refresher.ready().
  // start() only waited for the listener; converting before this line would ask
  // an empty book and get "unknown or unreachable currency pair" for every pair.
  const readyTimeoutMs = Number(process.env.OPENRATE_READY_TIMEOUT_MS ?? 20_000);
  let currencies = 0;
  try {
    currencies = await side.waitForRates(readyTimeoutMs);
  } catch (e) {
    // The message carries the 503's reason and each source's last error, so
    // this says WHICH source failed and how — not just "timed out".
    console.log(`rates       ${e instanceof Error ? e.message : "non-Error thrown"}`);
    console.log("            openrate says so rather than inventing a number; that is the design.");
    // And the exit code says it too. An example that prints a failure and exits
    // 0 is the same false green as converting against an empty book: whatever
    // runs this in CI sees success.
    process.exitCode = 1;
  }
  if (currencies > 0) {
    console.log(`rates       ${String(currencies)} currencies after the startup refresh`);

    const c = (await side.convert("EUR", "ZAR", 100)) as Convert;
    console.log(`convert     100 EUR = ${c.result.toFixed(2)} ZAR`);
    // The same audit trail direct mode returns — identical JSON, different
    // transport. That is the point of reusing the wire contract.
    console.log(`            path ${c.rate.path.join(" -> ")}, ${String(c.rate.hops)} hops, sources ${c.rate.sources.join(",")}`);
    console.log(`            ${(c.rate.age_sec / 3600).toFixed(1)}h old, grade ${c.rate.quality.grade}`);

    const book = (await side.rates("USD")) as { base: string; rates: Record<string, unknown> };
    console.log(`rates       base ${book.base} -> ${String(Object.keys(book.rates).length)} pairs`);
  }

  const meta = (await side.meta()) as { default_base: string; sources: { name: string; edges?: number }[] };
  console.log(`meta        default base ${meta.default_base}, sources ${JSON.stringify(meta.sources)}`);

  // The error path. Unlike the library, the HTTP endpoint answers 200 with an
  // empty book for an unknown base — but an unknown *pair* is still an error.
  try {
    await side.convert("XXX", "ZAR", 1);
    console.log("error       UNEXPECTED: an unknown currency converted");
  } catch (e) {
    console.log(`error       ${e instanceof Error ? e.message : "non-Error thrown"}`);
  }
} finally {
  side.stop();
}

File: sdks/rust/examples/sidecar.rs
//! openrate as a supervised child process, over HTTP.
//!
//! ```text
//! go build -o /tmp/openrate ./cmd/openrate
//! OPENRATE_BINARY=/tmp/openrate cargo run --example sidecar
//! OPENRATE_BINARY=/tmp/openrate cargo run --example sidecar -- 5   # 5s readiness budget
//! ```
//!
//! The optional argument is how long to wait for rates. It is there so the
//! failure path can be seen without waiting 45 seconds for it:
//!
//! ```text
//! HTTPS_PROXY=http://127.0.0.1:1 HTTP_PROXY=http://127.0.0.1:1 \
//!   OPENRATE_BINARY=/tmp/openrate cargo run --example sidecar -- 5
//! ```
//!
//! Compare with `examples/direct.rs`, which is the better default for a Rust
//! program. The sidecar earns its place when several processes should share one
//! refresher, when you want the HTTP shell's rate limiter and CORS policy, or
//! when you are on a platform openrate has no prebuilt library for — which, for
//! openrate, is most of them. See the crate README.

use std::process::ExitCode;
use std::time::{Duration, Instant};

use openrate::sidecar::{Options, Sidecar};

fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("error: {e}");
            ExitCode::FAILURE
        }
    }
}

fn run() -> Result<(), Box<dyn std::error::Error>> {
    // How long to wait for rates. Short values are for demonstrating the
    // failure path; 45s is the honest budget for a cold ECB fetch.
    let ready_timeout = match std::env::args().nth(1) {
        Some(s) => Duration::from_secs(s.parse()?),
        None => Duration::from_secs(45),
    };

    // Spawns the child on a free loopback port and blocks until it is
    // LISTENING. On failure it kills what it started, so this `?` cannot leave
    // a server behind.
    let sc = Sidecar::start(Options {
        base: "EUR".into(),
        sources: Some("ecb".into()),
        ..Default::default()
    })?;
    // From here the process is owned by `sc` and stopped by Drop — on the
    // early returns below and on a panic, not only at the end. Rust has no
    // `defer`, so this is where the equivalent guarantee comes from.
    println!("sidecar:   {}", sc.base_url);

    // ---------------------------------------------------------------- ready
    // Liveness first: this is already true, and it means almost nothing.
    println!(
        "healthz:   {:?}  (liveness only — no rates implied)",
        sc.healthz()?
    );

    // Readiness is a different question. /healthz answered `ok` before any
    // source had been fetched; wait_ready polls /readyz, which is a 503 with a
    // reason until the engine actually holds rates and a 200 once it does. If
    // this times out the error names the source that failed and why — run it
    // behind a dead proxy to see that.
    let t = Instant::now();
    let ready = sc.wait_ready(ready_timeout)?;
    println!("ready:     after {:?}", t.elapsed());
    println!(
        "readyz:    {}",
        first(&openrate::sidecar::compact(&ready), 200)
    );

    // /readyz counts currencies and does not name the presentation base, so
    // meta is still the place to ask for that. It is an API call now, not a
    // readiness probe.
    let meta = sc.meta()?;
    println!(
        "meta:      {}",
        first(&openrate::sidecar::compact(&meta), 200)
    );

    // -------------------------------------------------------------- convert
    // The same document the C ABI's "convert" method returns, over a different
    // transport. One wire contract.
    let conv = sc.convert("EUR", "USD", 100.0)?;
    println!("EUR->USD:  {}", summarize(&conv));

    // ---------------------------------------------------------------- rates
    let rates = sc.rates("EUR")?;
    println!("rates EUR: {} bytes", rates.len());

    // The alignment, demonstrated rather than described. This request is an
    // error over the C ABI and a 404 here, and both carry the same sentence —
    // it used to be a 200 with an empty book, which a client checking only the
    // status code read as success.
    match sc.rates("XXX") {
        Ok(v) => println!(
            "rates XXX: UNEXPECTEDLY answered: {}",
            first(&openrate::sidecar::compact(&v), 80)
        ),
        Err(e) => println!(
            "rates XXX: {}   (the C ABI returns the same \"unknown base currency\")",
            first(&e.to_string(), 160)
        ),
    }

    // ----------------------------------------------------------- error path
    // A pair the server cannot reach is a 404 with a JSON error body, and the
    // HTTP client surfaces that body verbatim rather than paraphrasing a code.
    match sc.convert("XXX", "YYY", 1.0) {
        Ok(v) => println!("bogus:     unexpectedly answered: {}", first(&v, 120)),
        Err(e) => println!("bogus:     {}", first(&e.to_string(), 160)),
    }

    println!("stopping:  Drop kills and reaps the child");
    Ok(())
}

fn summarize(json: &str) -> String {
    // Compact first: the HTTP API pretty-prints, so `"result": 115.35` has a
    // space the C ABI's output does not. Same nesting caveat as the direct
    // example too — `result` is top level, but `rate`, `hops` and `grade` live
    // inside the `"rate"` object.
    let json = &openrate::sidecar::compact(json);
    let result = num(json, "\"result\":");
    let rate = num(json, "\"rate\":{\"rate\":");
    let hops = num(json, "\"hops\":");
    let grade = str_field(json, "\"grade\":\"");
    format!("result={result} rate={rate} hops={hops} grade={grade}")
}

fn num(json: &str, key: &str) -> String {
    let Some(i) = json.find(key) else {
        return "?".into();
    };
    json[i + key.len()..]
        .chars()
        .take_while(|c| c.is_ascii_digit() || *c == '.' || *c == '-' || *c == 'e' || *c == '+')
        .collect::<String>()
}

fn str_field(json: &str, key: &str) -> String {
    let Some(i) = json.find(key) else {
        return "?".into();
    };
    json[i + key.len()..]
        .chars()
        .take_while(|c| *c != '"')
        .collect()
}

fn first(s: &str, max: usize) -> String {
    if s.chars().count() <= max {
        return s.to_string();
    }
    let t: String = s.chars().take(max).collect();
    format!("{t}…")
}
